### ISSUE TITLE
QUERY_ALL_PACKAGES

### ISSUE THREAD

**Issue by IzzySoft** [Mahmud0808/ColorBlendr#15]
Just wondering: from the app's description and screenshots, theming is applied globally. Still, [the app requests](https://github.com/Mahmud0808/ColorBlendr/blob/master/app/src/main/AndroidManifest.xml#L20) the `android.permission.QUERY_ALL_PACKAGES`. May I ask what for that is needed? Thanks for clarification!

**Comment by Mahmud0808** [comment]
It is because in the latest material components library v1.11.0+, apps don't follow system colors anymore. They now need to be individually themed. So ColorBlendr shows the list of all installed apps and if user has some apps which are monet themed but still not following system colors, they will need to select it and it will be themed instantly.

Related explanation (shown in-app too): [strings.xml#L114](https://github.com/Mahmud0808/ColorBlendr/blob/master/app/src/main/res/values/strings.xml#L114)
Related issues (resolved): https://github.com/Mahmud0808/ColorBlendr/issues/5

**Comment by IzzySoft** [comment]
Ah, thanks! May I suggest to point this out in the app description as well (not necessarily the permission itself, but at least that it needs to access and show a list of all apps) so the question won't resurface every once? Some people prefer to know *before* installing, so they won't see it "in-app". Even providing a screenshot of it showing the apps would do, though then is the risk nobody remembers why this specific screenshot is needed and it might get replaced with something else one day.

**Comment by IzzySoft** [comment]
PS, that clarified: the app really needs to access **ALL** storage, like a file manager does? Asking because of `android.permission.MANAGE_EXTERNAL_STORAGE`.

**Comment by Mahmud0808** [comment]
> May I suggest to point this out in the app description as well

README or fastlane project description?

> the app really needs to access ALL storage, like a file manager does?

Yes. To generate colors from wallpaper, it needs to grab the wallpaper first. And without this permission it will throw error.

**Comment by IzzySoft** [comment]
> README or fastlane project description?

At least Fastlane (which is what will be shown in the F-Droid clients later). README cannot hurt either, for everyone else. Speaking of Fastlane and F-Droid: You might wish to add a [FUNDING.yml](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/displaying-a-sponsor-button-in-your-repository) to your `.github` directory for your donation links. F-Droid will pick that up along with Fastlane.

And also btw Fastlane: featureGraphic dimensions should be either 1024x500 or 512x250, yours is 4096x1440 and thus does not even have the correct proportions (it also gives another warning: "libpng warning: iCCP: known incorrect sRGB profile"). Not a real stopper, but might look weird in some places.

(storage) Thanks! Then all should be fine. Not just for your RFP, your app just passed all other tests here for my repo, too, and [will show up in my repo](https://apt.izzysoft.de/packages/com.drdisagree.colorblendr/) with the next sync in less than two hours from now.

Feel free to close this issue if you think all has been addressed – or do so after adjusting featureGraphic and `FUNDING.yml` if you prefer :wink: Thanks a lot!

**Comment by Mahmud0808** [comment]
I will close this issue after resolving all and also will try to find alternative permissions for getting wallpaper (not sure if there is any). Thanks for the help btw.

**Comment by IzzySoft** [comment]
I've put those two permissions on the positive-list (so scans on future updates won't raise alerts here). F-Droid itself (where you applied for inclusion) does not scan for those on updates (yet), only on inclusion (I've sent a note with the details to LK, so the team is aware this was already addressed).

Thanks for updating Fastlane details! Is there any new release scheduled soon, or should I manually trigger a fastlane update here so it gets into my repo? On new releases it will be pulled along automatically.

**Comment by Mahmud0808** [comment]
There is no new release scheduled for now. Thanks for the heads-up.

**Comment by IzzySoft** [comment]
OK, just triggered a fastlane pull. Your update will show up then with the next sync around 7 pm UTC. Thanks a lot!

**Comment by IzzySoft** [comment]
Today it comes up with `android.permission.READ_PHONE_STATE` 🤔  What would that be needed for, @Mahmud0808? Release notes don't tell.

Oh, btw: isn't there a badge missing in the Readme? See our [Badges & Shields](https://izzyondroid.org/about/resources/badges/) and take your pick if you like 😃

**Comment by Mahmud0808** [comment]
`android.permission.READ_PHONE_STATE` This permission is not directly used by ColorBlendr. This is from AndroidX WorkManager library which is used for running background tasks.

**Comment by IzzySoft** [comment]
That makes no sense to me. Why should background tasks require phone state (i.e. if there's an ongoing call, with what number, what provider is serving you, the device's IMEI & serial (thus having a unique identifier!), etc)? As the app also requests `INTERNET`, and one cannot see any connection to telephony functionality in the app, this _at least_ raises eye brows, if not concerns.

We have almost 1.300 apps in our repo. I guess yours is not the only one using AndroidX Workmanager – but I cannot remember having heard this before (though something nags at the back of my brains). I did a quick web search for `"AndroidX Workmanager" +"READ_PHONE_STATE"`, but with most search engines this did not turn up a single result (with others, max 1..3 – but none really fitting). So are you sure with this? Maybe you can try removing this permission, and see whether it has any impact (and should it have, which)? See e.g. [Removing Unwanted Manifest Permissions With tools:node](https://developermemos.com/posts/removing-unwanted-manifest-permissions/).

OK, did another search, this time leaving out the quotes around "AndroidX Workmanager". Found something, e.g. [here at Reddit](https://www.reddit.com/r/androiddev/comments/eyr2o5/deleted_by_user/) (funnily, the only matching result still). There seem to be rare cases where WorkManager does this. And again, the recommendation is that you can use `tools:node` to remove it then, as it seems to be unused anyway.

**Comment by Mahmud0808** [comment]
Oh wait my bad, I thought we are talking about `android.permission.ACCESS_NETWORK_STATE`.

`android.permission.READ_PHONE_STATE` I can't find this anywhere in my source code. Not even in merged manifest. But I see it's there in the APK. Strange to me too.

It's probably injected by a dependency but Android Studio is not showing which one. And I am not sure if `tools:node` will be a good choice because it might break that dependency functionality.

**Comment by IzzySoft** [comment]
`ACCESS_NETWORK_STATE` goes well with `INTERNET` (a la "first make sure we can"). And as for `tools:node` on `READ_PHONE_STATE`: maybe try to see? Won't know for sure otherwise. And the 2 cases I found on this, didn't state it broke anything. But yeah, 2 cases are not many – but still, worth a try?

**Comment by Mahmud0808** [comment]
[This](https://commonsware.com/blog/2015/06/25/hey-where-did-these-permissions-come-from.html) helped me find out where the permission is coming from.
TL;DR: Open `app/build/outputs/logs/manifest-merger-debug-report.txt` and search for the permission.

Edit: Released a new build. Now it doesn't contain `android.permission.READ_PHONE_STATE`. Thanks for the report.

**Comment by IzzySoft** [comment]
> [This](https://commonsware.com/blog/2015/06/25/hey-where-did-these-permissions-come-from.html) helped me find out where the permission is coming from.

Wow, thanks for sharing! [Adding that](https://codeberg.org/IzzyOnDroid/website/pulls/71) right to our documentation 🤗

> Released a new build. Now it doesn't contain `android.permission.READ_PHONE_STATE`.

Manually triggered an update: confirmed, that permission is gone, thanks! Also, RB confirmed for this one.

> Thanks for the report.

Thank you for taking care – and sharing those details!!! 🤗
